File: mobile/ios/AutoRL/MainViewModel.swift
//
//  MainViewModel.swift
//  AutoRL
//
//  ViewModel for MainView following MVVM architecture
//  Manages model loading, inference execution, and UI state
//

import Foundation
import Combine
import SwiftUI

class MainViewModel: ObservableObject {
    // Published properties for SwiftUI binding
    @Published var statusText: String = "Ready to start"
    @Published var isLoading: Bool = false
    @Published var resultText: String = ""
    @Published var latencyText: String = ""
    @Published var isStartButtonEnabled: Bool = false
    @Published var showRetryButton: Bool = false
    @Published var errorMessage: String? = nil
    @Published var successMessage: String? = nil
    
    // Dependencies
    private let modelRunner: ModelRunner
    private let apiService: APIService
    
    // State management
    private var cancellables = Set<AnyCancellable>()
    
    enum UIState {
        case idle
        case loading
        case modelLoaded
        case runningInference
        case inferenceComplete(result: String, latencyMs: Int64)
        case error(message: String)
    }
    
    private var currentState: UIState = .idle {
        didSet {
            updateUIForState(currentState)
        }
    }
    
    init(modelRunner: ModelRunner = ModelRunner.shared,
         apiService: APIService = APIService.shared) {
        self.modelRunner = modelRunner
        self.apiService = apiService
    }
    
    /// Initialize and load the model
    func initializeModel() {
        Task { @MainActor in
            do {
                currentState = .loading
                
                let success = await modelRunner.loadModel()
                
                if success {
                    currentState = .modelLoaded
                    successMessage = "Model loaded successfully"
                    print("✅ Model loaded successfully")
                } else {
                    currentState = .error(message: "Failed to load model")
                    errorMessage = "Could not load the AI model. Please check if the model file exists."
                }
            } catch {
                let errorMsg = "Error loading model: \(error.localizedDescription)"
                currentState = .error(message: errorMsg)
                errorMessage = errorMsg
                print("❌ Failed to load model: \(error)")
            }
        }
    }
    
    /// Run inference on a test image or screenshot
    func runInference() {
        Task { @MainActor in
            do {
                guard modelRunner.isModelLoaded else {
                    errorMessage = "Model not loaded. Please wait for model to load."
                    return
                }
                
                currentState = .runningInference
                
                let result = await modelRunner.runInference()
                
                if let result = result {
                    currentState = .inferenceComplete(
                        result: result.output,
                        latencyMs: result.latencyMs
                    )
                    successMessage = "Inference completed successfully"
                    print("✅ Inference completed in \(result.latencyMs)ms")
                } else {
                    currentState = .error(message: "Inference returned no result")
                    errorMessage = "Inference failed. Please try again."
                }
            } catch {
                let errorMsg = "Error during inference: \(error.localizedDescription)"
                currentState = .error(message: errorMsg)
                errorMessage = errorMsg
                print("❌ Inference failed: \(error)")
            }
        }
    }
    
    /// Dismiss error message
    func dismissError() {
        errorMessage = nil
    }
    
    /// Dismiss success message
    func dismissSuccess() {
        successMessage = nil
    }
    
    /// Update UI based on current state
    private func updateUIForState(_ state: UIState) {
        switch state {
        case .idle:
            statusText = "Ready to start"
            isLoading = false
            isStartButtonEnabled = false
            showRetryButton = false
            resultText = ""
            latencyText = ""
            
        case .loading:
            statusText = "Loading model..."
            isLoading = true
            isStartButtonEnabled = false
            showRetryButton = false
            
        case .modelLoaded:
            statusText = "Model loaded"
            isLoading = false
            isStartButtonEnabled = true
            showRetryButton = false
            
        case .runningInference:
            statusText = "Running inference..."
            isLoading = true
            isStartButtonEnabled = false
            showRetryButton = false
            
        case .inferenceComplete(let result, let latencyMs):
            statusText = "Inference complete"
            isLoading = false
            isStartButtonEnabled = true
            showRetryButton = false
            resultText = result
            latencyText = String(format: "Latency: %lld ms", latencyMs)
            
        case .error(let message):
            statusText = message
            isLoading = false
            isStartButtonEnabled = false
            showRetryButton = true
            resultText = ""
            latencyText = ""
        }
    }
    
    deinit {
        Task {
            await modelRunner.cleanup()
        }
    }
}

File: mobile/android/app/src/main/java/com/autorl/MainViewModel.kt
package com.autorl

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import timber.log.Timber
import javax.inject.Inject

/**
 * ViewModel for MainActivity following MVVM architecture.
 * 
 * Manages:
 * - Model loading state
 * - Inference execution
 * - UI state updates
 * - Error handling
 */
@HiltViewModel
class MainViewModel @Inject constructor(
    private val modelRunner: ModelRunner
) : ViewModel() {
    
    sealed class UIState {
        object Idle : UIState()
        object Loading : UIState()
        object ModelLoaded : UIState()
        object RunningInference : UIState()
        data class InferenceComplete(
            val result: String,
            val latencyMs: Long
        ) : UIState()
        data class Error(val message: String) : UIState()
    }
    
    private val _uiState = MutableStateFlow<UIState>(UIState.Idle)
    val uiState: StateFlow<UIState> = _uiState.asStateFlow()
    
    private val _errorMessage = MutableStateFlow<String?>(null)
    val errorMessage: StateFlow<String?> = _errorMessage.asStateFlow()
    
    private val _successMessage = MutableStateFlow<String?>(null)
    val successMessage: StateFlow<String?> = _successMessage.asStateFlow()
    
    init {
        Timber.d("MainViewModel initialized")
    }
    
    /**
     * Initialize and load the model.
     */
    fun initializeModel() {
        viewModelScope.launch {
            try {
                _uiState.value = UIState.Loading
                
                val success = modelRunner.loadModel()
                
                if (success) {
                    _uiState.value = UIState.ModelLoaded
                    _successMessage.value = "Model loaded successfully"
                    Timber.d("Model loaded successfully")
                } else {
                    _uiState.value = UIState.Error("Failed to load model")
                    _errorMessage.value = "Could not load the AI model. Please check if the model file exists."
                }
            } catch (e: Exception) {
                val errorMsg = "Error loading model: ${e.message}"
                _uiState.value = UIState.Error(errorMsg)
                _errorMessage.value = errorMsg
                Timber.e(e, "Failed to load model")
            }
        }
    }
    
    /**
     * Run inference on a test image or bitmap.
     */
    fun runInference() {
        viewModelScope.launch {
            try {
                if (!modelRunner.isModelLoaded()) {
                    _errorMessage.value = "Model not loaded. Please wait for model to load."
                    return@launch
                }
                
                _uiState.value = UIState.RunningInference
                
                val result = modelRunner.runInference()
                
                if (result != null) {
                    _uiState.value = UIState.InferenceComplete(
                        result = result.output,
                        latencyMs = result.latencyMs
                    )
                    _successMessage.value = "Inference completed successfully"
                    Timber.d("Inference completed in ${result.latencyMs}ms")
                } else {
                    _uiState.value = UIState.Error("Inference returned no result")
                    _errorMessage.value = "Inference failed. Please try again."
                }
            } catch (e: Exception) {
                val errorMsg = "Error during inference: ${e.message}"
                _uiState.value = UIState.Error(errorMsg)
                _errorMessage.value = errorMsg
                Timber.e(e, "Inference failed")
            }
        }
    }
    
    /**
     * Mark error message as shown.
     */
    fun errorMessageShown() {
        _errorMessage.value = null
    }
    
    /**
     * Mark success message as shown.
     */
    fun successMessageShown() {
        _successMessage.value = null
    }
    
    /**
     * Cleanup resources.
     */
    fun cleanup() {
        viewModelScope.launch {
            modelRunner.cleanup()
        }
    }
    
    override fun onCleared() {
        super.onCleared()
        cleanup()
        Timber.d("MainViewModel cleared")
    }
}

